File: packages/study/agent-ts-blocks/src/engine.ts
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { GoogleGenAI } from '@google/genai';
import type { ChatMessage, AgentComponent } from './types.js';

/**
 * The core industrial agent loop.
 * Orchestrates components and manages the LLM lifecycle.
 */
export class AgentEngine {
  private components: AgentComponent[] = [];
  private ai: GoogleGenAI;
  private modelName: string;

  constructor(apiKey: string, modelName = 'gemini-3-flash-preview') {
    this.ai = new GoogleGenAI({ apiKey });
    this.modelName = modelName;
  }

  /**
   * Registers a pluggable component.
   */
  use(component: AgentComponent): this {
    this.components.push(component);
    return this;
  }

  /**
   * Executes a single chat turn.
   */
  async *chat(userInput: string): AsyncGenerator<string> {
    let messages: ChatMessage[] = [{ role: 'user', content: userInput }];

    // 1. Run 'beforeLLMCall' hooks (e.g., Memory injection)
    for (const component of this.components) {
      if (component.beforeLLMCall) {
        messages = await component.beforeLLMCall(messages);
      }
    }

    // 2. Prepare for Gemini API
    // Note: Gemini uses 'user' and 'model' as roles
    const contents = messages.map((m) => ({
      role: m.role,
      parts: [{ text: m.content }],
    }));

    // 3. Call LLM
    const result = await this.ai.models.generateContentStream({
      model: this.modelName,
      contents,
    });

    // 4. Stream response and capture full text
    let fullResponse = '';
    for await (const chunk of result) {
      const text = chunk.text || '';
      fullResponse += text;
      yield text;
    }

    // 5. Run 'afterLLMCall' hooks (e.g., Save to memory)
    for (const component of this.components) {
      if (component.afterLLMCall) {
        await component.afterLLMCall(userInput, fullResponse);
      }
    }
  }
}

File: packages/study/agent-py-blocks/src/engine.py
from typing import List, AsyncGenerator
from google import genai
from .types import ChatMessage, AgentComponent

class AgentEngine:
    def __init__(self, api_key: str, model_name: str = "gemini-3-flash-preview"):
        self.client = genai.Client(api_key=api_key)
        self.model_name = model_name
        self.components: List[AgentComponent] = []

    def use(self, component: AgentComponent) -> "AgentEngine":
        self.components.append(component)
        return self

    async def chat(self, user_input: str) -> AsyncGenerator[str, None]:
        messages = [ChatMessage(role="user", content=user_input)]

        # 1. Run before_llm_call hooks
        for component in self.components:
            if hasattr(component, "before_llm_call"):
                messages = await component.before_llm_call(messages)

        # 2. Prepare contents for API
        contents = [{"role": m.role, "parts": [{"text": m.content}]} for m in messages]

        # 3. Call LLM
        response_stream = self.client.models.generate_content_stream(
            model=self.model_name,
            contents=contents
        )

        # 4. Stream and capture
        full_response = ""
        async for chunk in response_stream:
            text = chunk.text or ""
            full_response += text
            yield text

        # 5. Run after_llm_call hooks
        for component in self.components:
            if hasattr(component, "after_llm_call"):
                await component.after_llm_call(user_input, full_response)
